### PR TITLE
Feature: Weight adjustment utility

### LINKED ISSUE #22: Feature: Weight normalization & interpolation
## Status Quo
- on multiple occasions in the code we need to normalize & interpolate weights

## Acceptance Criteria
- there is a helper function `normalizeInterpolated`
- it takes as input params:
  - `baseWeights`: the initial weights for all tokens that are in the pool or that should be added to the pool (e.g. for a two-token pool with weights 80/20 to which we want to add a third token this would be 80/20/0)
  - `targetWeights`: the target weights for the tokens that we want to set to a fixed value, the values for all other tokens are provided as zeros (e.g. we want to set the weight of the new token to 1%, this would be 0/0/1)
- it returns the interpolated normalized weights (in the example this would be 79.2/19.8/1)

## Additional info
- per our latest discussion (29/10/2021) it will not be possible to set the weight of a token to zero

### PR DESCRIPTION
**Pull request by fabianschu**
I think it might be a good idea to use the newly created contract in this PR to serve as a location for utility functions for the IndexPool, from which the IndexPool could inherit. Wdyt?

Since this code is a bit critical for the whole pool, I added tests with randomized inputs. Most of the test cases are basically the same, just with different token amounts :)

Lastly, I used the `TestWrapper.sol` contract solely for the purpose of being able to call the internal function `normalizeInterpolated`

closes #22

### CODE REVIEW

File: pkg/pool-weighted/test/IndexPoolUtils.test.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+        beforeEach(async () => {
+          expectedWeights = getExpectedWeights(baseWeights, fixedWeights);
+          receivedWeights = await normalizerInstance.normalizeInterpolated(
+            baseWeights.map((w) => toBnPercentage(w)),
+            fixedWeights.map((w) => toBnPercentage(w))
+          );
+        });
+
+        it('returns the correct weights', async () => {
```
[the remark this answers is not in the document]
Comment (fabianschu) [reply pkg/pool-weighted/test/IndexPoolUtils.test.ts]: I add this case, but afaiu new tokens will always initially be added with a weight of 0,01, so I can't think of a case where your described case should ever occur.

File: pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol [hunk 2]
```diff
@@ -0,0 +1,75 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.7.0;
+
+contract IndexPoolUtils {
+    uint8 public constant PRECISION = 18;
+    uint256 public constant BONE = 10**PRECISION;
```
[the remark this answers is not in the document]
Comment (fabianschu) [reply pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol]: Ah, that's how balancer v1 called HUNDRED_PERCENT. I changed the name. It was weird 🦴

**Comment by fabianschu** [comment]
@BritikovKI Since your last review I added randomized tests which helped me to spot some edge cases where the returned weights would not be normalized. To catch that I inserted some additional logic [here](https://github.com/PrimeDAO/balancer-v2-monorepo/blob/research/weight-adjustment/pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol#L71)

File: pkg/pool-weighted/test/IndexPoolUtils.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BritikovKI) [reply pkg/pool-weighted/test/IndexPoolUtils.test.ts]: Yep, I understand that this case won't actually be met in life, I asked this just to check the math correctness

File: pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (BritikovKI) [reply pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol]: LUL, understood

**Review by BritikovKI — APPROVED** [review]
Wonderful, I like tests with numbers a lot more :)

**Review by Zitzak — APPROVED** [review]
